### ISSUE TITLE
Emoji placeholders still visible after migration

### ISSUE THREAD

**Issue by anx-ag** [ergon/migratemost#16]
Hey,

I did run migratemost using the flag to include custom emojis and it didn't throw any errors, also I can see that the custom emojis are all here now (:hc_*), but in the rooms/channels, I can see that not all occurences of emojis have been replaced, f.ex.:

(y) shows as (y) when it should be a thumbsup icon.
(thumbsup) f.ex. is also one of the hipchat integrated emojis, so I'm not even talking about custom emojis here, they don't work either.

Is that an expected behaviour?

Thanks,
Alex

**Comment by anx-ag** [comment]
Here's a patch that fixes this problem (it also includes the fix for the newline issue in #18 ):

```
--- migratemost.py.orig 2019-09-20 14:53:50.422237709 +0200
+++ migratemost.py      2019-09-20 16:46:20.672943073 +0200
@@ -406,25 +408,49 @@
     return int(time.mktime(d.timetuple())) * 1000 + d.microsecond / 1000


+_hc_emoji_cache = None
 def sanitize_message(message):
     # Translate Hipchat formatting to Mattermost and split too long messages
     # List of slash commands in Hipchat:
     # https://confluence.atlassian.com/hipchatdc3/keyboard-shortcuts-and-slash-commands-966656108.html
+
+    global _hc_emoji_cache
+    global option_hipchat_base_url
+    global option_hipchat_tokens
+    if _hc_emoji_cache is None:
+        from migrate_hipchat_emoticons import _fetch_emoticons
+        _hc_emoji_cache = _fetch_emoticons(option_hipchat_base_url, option_hipchat_tokens)
+
+    translation = {
+        "(thumbsup)": ":+1:",
+        "(thumbsdown)": ":-1:",
+        "(oops)": ":sweat:",
+        "(embarrassed)": ":flushed:",
+    }
+    for hc_emoticon in _hc_emoji_cache:
+        hc_emoji = '(%s)' % (hc_emoticon[u'shortcut'],)
+        mm_emoji = ':hc_%s:' % (hc_emoticon[u'shortcut'],)
+        translation[hc_emoji] = mm_emoji
+    def _trans(s, t):
+        for a, b in t.items():
+            s = s.replace(a, b)
+        return s
+
     message_parts = ['']
     if message.startswith("/code"):
-        sliced = textwrap.wrap(message[6:],
-                               MM_MAX_MESSAGE_LENGTH - 8)  # shorten 8 to make room for formatting characters
-        message_parts = ["```\n%s\n```" % m for m in sliced]
+        message_parts = ["```\n%s\n```" % (message[6:], )]
     elif message.startswith("/quote"):
+        message = _trans(message, translation)
+
         sliced = textwrap.wrap(message[7:],
                                MM_MAX_MESSAGE_LENGTH - 3)  # shorten 3 to make room for formatting characters
         message_parts = ["> %s\n" % m for m in sliced]
     else:
+        message = _trans(message, translation)
+
         message_parts = textwrap.wrap(message, MM_MAX_MESSAGE_LENGTH)
```

**Comment by 7-plus-t** [comment]
I've refactored your change a bit, as it is suboptimal to make this call in the `sanitize_message` method. It is called **a lot** and this would most likely have quite a bad impact on performance. I've moved it and changed it slightly, see https://github.com/ergon/migratemost/commit/8bd342a905c6dabebe32f94bfdcb36e62d19c5a3

Thanks for adding this feature!
